File: Исходники/Глава 2. Часть 1/Жадные алгоритмы/7. Алгоритм Прима/Python/MST/Prima_Olimp.py

edges = [(2, 5, 1), (4, 3, 5), (0, 1, 11), (2, 0, 7), (1, 3, 5), (4, 2, 2), (5, 3, 3)]
vtx = 6


def min_edge(visit, edges):
    arr = []
    for edge in edges:
        if edge[0] in visit or edge[1] in visit:
            if edge[0] not in visit or edge[1] not in visit:
                arr.append(edge)
    min_ = min(arr, key=lambda x: x[2])
    visit.add(min_[1])
    visit.add(min_[0])
    return min_


def prima(edges):
    mst = 0
    visit = {0}
    while len(visit) != vtx:
        mst += min_edge(visit, edges)[2]
    return mst


print(prima(edges))

File: Исходники/Глава 3. Часть 1/33. Минимальное остовное дерево/C++/Олимпиадный/Prima_Olimp.cpp
#include <iostream>
#include <set>
#include <vector>
#include <algorithm>

using namespace std;

typedef vector<int> vi;
typedef set<int> si;

bool comp(vi edge1, vi edge2) {
    return edge1[2] < edge2[2];
}

vi r_min_edge(vector<vi> edges, si visit) {
    vector<vi> arr;
    for (vi edge : edges) {
        if ((visit.count(edge[0]) != 0) || (visit.count(edge[1]) != 0)) {
            if ((visit.count(edge[0])) == 0 || (visit.count(edge[1])) == 0) {
                arr.push_back(edge);
            }
        }
    };
    sort(arr.begin(), arr.end(), comp);
    vi min_edge = arr[0];
    arr.clear();
    return min_edge;
}

int prima(vector<vi> edges, si visit, int vtx) {
    int mst = 0;
    while (visit.size() != vtx) {
        vi min_edge = r_min_edge(edges, visit);
        mst += min_edge[2];
        visit.insert(min_edge[0]);
        visit.insert(min_edge[1]);
    };
    return mst;
}

int main() {
    vector<vi> edges = { {2, 5, 1}, {4, 3, 5}, {0, 1, 11}, {2, 0, 7}, {1, 3, 5}, {4, 2, 2}, {5, 3, 3} };
    si visit = { 0 };
    int vtx = 6;
    cout << prima(edges, visit, vtx);
    return 0;
}
